src/main/java/frc/robot/subsystems/swerve/SwerveSubsystem.java

```diff
@@ -554,16 +562,32 @@ private void runVelocity(ChassisSpeeds speeds) {
         ChassisSpeeds.fromRobotRelativeSpeeds(discreteSpeeds, getRotation()));
 
     // Send setpoints to modules
-    SwerveModuleState[] optimizedSetpointStates =
-        Streams.zip(
-                Arrays.stream(modules), Arrays.stream(setpointStates), (m, s) -> m.runSetpoint(s))
-            .toArray(SwerveModuleState[]::new);
+    SwerveModuleState[] optimizedSetpointStates = new SwerveModuleState[modules.length];
+    SwerveModuleState[] forceSetpoints = new SwerveModuleState[modules.length];
+    for (int i = 0; i < optimizedSetpointStates.length; i++) {
+      var robotRelForceX =
+          moduleForcesX[i] * getRotation().getCos() - moduleForcesY[i] * getRotation().getSin();
+      var robotRelForceY =
+          moduleForcesX[i] * getRotation().getSin() + moduleForcesY[i] * getRotation().getCos();
+      forceSetpoints[i] =
+          new SwerveModuleState(
+              Math.hypot(moduleForcesX[i], moduleForcesY[i]),
+              new Rotation2d(moduleForcesX[i], moduleForcesY[i]));
+      optimizedSetpointStates[i] =
+          modules[i].runSetpoint(setpointStates[i], robotRelForceX, robotRelForceY);
+    }
 
     // Log setpoint states
     Logger.recordOutput("SwerveStates/Setpoints", setpointStates);
+    Logger.recordOutput("SwerveStates/ForceSetpoints", forceSetpoints);
     Logger.recordOutput("SwerveStates/SetpointsOptimized", optimizedSetpointStates);
   }
 
+  private void runVelocity(ChassisSpeeds speeds) {
+    // do the allocations here matter?
```

Comment:
If a 64 byte allocation is a problem, we're doing something right everywhere else.